gmso/formats/json.py

```diff
@@ -0,0 +1,308 @@
+from gmso.core.improper import Improper
+from gmso.core.improper_type import ImproperType
+from gmso.core.pairpotential_type import PairPotentialType
+from gmso.formats.formats_registry import loads_as, saves_as
+
+
+def _to_json(top, types=False, update=True):
+    """Return a json serializable dictionary from a topology.
+
+    This method is used for json serializing the topology
+
+    Parameters
+    ----------
+    top: gmso.Topology, required
+        The topology
+    types: bool, default=False
+        If true, include type info (i.e. Potentials)
+    update: bool, default=False
+        If true, update the topology before iterating through the files
+
+    Returns
+    -------
+    dict
+        A json serializable dictionary representing members of this Topology
+    """
+    if types and not top.is_typed():
+        raise ValueError(
+            "Cannot incorporate types because the topology is not typed."
+        )
+    if update:
+        top.update_topology()
+
+    json_dict = {
+        "name": top._name,
+        "scaling_factors": top.scaling_factors,
+        "subtopologies": [],
+        "box": top.box.json_dict() if top.box else None,
+        "atoms": [],
+        "bonds": [],
+        "angles": [],
+        "dihedrals": [],
+        "impropers": [],
+        "atom_types": [],
+        "bond_types": [],
+        "angle_types": [],
+        "dihedral_types": [],
+        "improper_types": [],
+        "pair_potentialtypes": [],
+    }
+
+    for atom in top._sites:
+        atom_dict = atom.json_dict(exclude={"atom_type"})
+        if types and atom.atom_type:
+            atom_dict["atom_type"] = id(atom.atom_type)
+
+        json_dict["atoms"].append(atom_dict)
+
+    targets = {
+        Bond: json_dict["bonds"],
+        Angle: json_dict["angles"],
+        Dihedral: json_dict["dihedrals"],
+        Improper: json_dict["impropers"],
+        AtomType: json_dict["atom_types"],
+        BondType: json_dict["bond_types"],
+        AngleType: json_dict["angle_types"],
+        DihedralType: json_dict["dihedral_types"],
+        ImproperType: json_dict["improper_types"],
+    }
+
+    for connections, exclude_attr in [
+        (top._bonds, "bond_type"),
+        (top._angles, "angle_type"),
+        (top._dihedrals, "dihedral_type"),
+        (top._impropers, "improper_type"),
+    ]:
+        for connection in connections:
+            connection_dict = connection.json_dict(
+                exclude={exclude_attr, "connection_members"}
+            )
+            target = targets[type(connection)]
+            connection_dict["connection_members"] = [
+                top.get_index(member)
+                for member in connection.connection_members
+            ]
+            target.append(connection_dict)
+            connection_type = getattr(connection, exclude_attr)
+            if types and connection_type:
+                connection_dict[exclude_attr] = id(connection_type)
+    if types:
```

Comment:
- Partially parameterized systems should warn the users if they are saving it with `types=True`.
- Fully/Partially parameterized systems should warn the users that type info will be lost during conversion with `types=False`.

Reply:
This has been added